### PR TITLE
[BEAM-882,BEAM-883,BEAM-878] Simplified API surface verifications (take 2)

### PR DESCRIPTION
**Pull request by staslev**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by staslev** [comment]
This PR is pending the resolution of [beam_PostCommit_Java_RunnableOnService_Dataflow #2085](https://lists.apache.org/thread.html/67ef0a9ef4dc2ac33c2ccb997388b27bb47c58d92ec26a0a0e38cb5b@%3Cdev.beam.apache.org%3E) which does not seem to reproduce locally and is undergoing an investigation, see also #1822.

**Comment by staslev** [comment]
The issue can be reproduced locally by running:
 `mvn clean verify -am -pl runners/google-cloud-dataflow-java -DforkCount=0`,
setting `forkCount=1` makes this problem go away due to reasons detailed in [maven-failsafe-plugin and class loading](http://maven.apache.org/surefire/maven-failsafe-plugin/examples/class-loading.html).

Essentially the problem is related to how the `ClassLoader` that is used to query for `sdk` classes is obtained. At the moment it is obtained by using `ClassLoader.getSystemClassLoader()`, and in the case of post commit jobs maven runs with `forkCount=0`, and the system `ClassLoader` is not what one would expect. In particular, it does not have classes belonging to `org.apache.beam` present in its classpath.

My latest change should make things happy regardless of what `forkCount` is set to.

@kennknowles , let me know what you think.

**Comment by kennknowles** [comment]
I've forgotten why we use `forkCount=0`, but grabbing the right `ClassLoader` seems like the right move.

**Comment by staslev** [comment]
LGTY?

**Comment by staslev** [comment]
Travis failures seem unrelated.

**Comment by staslev** [comment]
retest this please

**Comment by kennknowles** [comment]
Just for the record, leveraging our prior review for anyone dropping in:

```bash
$ git diff --stat 29ffaf3859ba9b4d8ba8529efc96fd5e105e21a3..github/pr/1888 -- \
    sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java \
    sdks/java/core/src/test/java/org/apache/beam/SdkCoreApiSurfaceTest.java \
    sdks/java/core/src/test/java/org/apache/beam/sdk/util/ApiSurfaceTest.java \
    sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/ApiSurfaceTest.java \
    sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/GcpApiSurfaceTest.java 
```

yields

```
sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java
 | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)
```

and the tiny relevant diff is:

```
diff --git a/sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java b/sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java
index b6b0b3233..8a38d504f 100644
--- a/sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java
+++ b/sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java
@@ -322,6 +322,14 @@ public class ApiSurface {

   ///////////////

+  private ClassLoader getSdkClassLoader() {
+    final ClassLoader myClassLoader = getClass().getClassLoader();
+    final ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
+    return systemClassLoader.equals(myClassLoader)
+        ? systemClassLoader
+        : myClassLoader;
+  }
+
   /** Returns an empty {@link ApiSurface}. */
   public static ApiSurface empty() {
     LOG.debug("Returning an empty ApiSurface");
@@ -348,7 +356,7 @@ public class ApiSurface {
    * its subpackages.
    */
   public ApiSurface includingPackage(String packageName) throws IOException {
-    ClassPath classPath = ClassPath.from(ClassLoader.getSystemClassLoader());
+    ClassPath classPath = ClassPath.from(getSdkClassLoader());

     Set<Class<?>> newRootClasses = Sets.newHashSet();
     for (ClassInfo classInfo : classPath.getTopLevelClassesRecursive(packageName)) {
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java [hunk 1]
```diff
@@ -66,45 +80,283 @@
...
+      }
+    }
+  }
+
+  ///////////////
+
+  private ClassLoader getSdkClassLoader() {
+    final ClassLoader myClassLoader = getClass().getClassLoader();
+    final ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
+    return systemClassLoader.equals(myClassLoader)
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java]: This is equivalent to just `return myClassLoader`, isn't it?

I think the SDK's API surface should own providing the class loader for this. So I would get the class loader in `SdkCoreApiSurfaceTest` and pass it in to `ApiSurface`.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java [hunk 2]
```diff
@@ -66,45 +80,283 @@
...
 
   /**
-   * Returns an {@link ApiSurface} like this one, but also including the named
-   * package and all of its subpackages.
+   * Returns an {@link ApiSurface} like this one, but also including the named package and all of
+   * its subpackages.
    */
   public ApiSurface includingPackage(String packageName) throws IOException {
-    ClassPath classPath = ClassPath.from(ClassLoader.getSystemClassLoader());
+    ClassPath classPath = ClassPath.from(getSdkClassLoader());
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java]: I think the `ClassLoader` here should come from `this`.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java (hunk 1 above)
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java]: That's a good point, it appears I got a bit over-zealous to try and preserve previous behavior :)

I think that making the `ClassLoader` instance a dependency that is passed in by the caller makes a lot of sense.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java (hunk 2 above)
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java]: Sounds good.

**Comment by staslev** [comment]
Rebased against master and revised according to comments.

**Comment by kennknowles** [comment]
LGTM, just wait to see from the postcommit at https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Dataflow/2226/

**Comment by kennknowles** [comment]
(I think the other postcommits were already fine, right?)

**Comment by staslev** [comment]
[beam_PostCommit_Java_RunnableOnService_Dataflow/2226/](https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Dataflow/2226/) failed on seemingly unrelated errors. Looking into it...

**Comment by kennknowles** [comment]
Yea, those failures are just failures to upload files. Not indicative of anything.

**Comment by kennknowles** [comment]
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Dataflow/2231/ is a rebuild (when it gets through the queue)

**Comment by staslev** [comment]
I've also triggered one (#2230), cancelling.

**Pull request closed without merging**